### PR TITLE
migrate dmpro

### PR DESCRIPTION
**Pull request by ThyWoof**
- migrates all DMPRO features except for custom monsters and multiplayer
- **EnableDungeonMakerPro** is a master hidden switch in the rare case we wanna fully disable this from CE
- **EnableDungeonMakerModdedContent** controls if authors will have access to additional assets in the editor
- **WorldLocationPatcher** is only controlled by the master switch to keep the experience as bug free as possible to players (no need to toggle any switch. all a player requires is install CE to play modded dungeon content)
- custom content in the editor is yellow and tagged as [modded]
- added proper text explaining the consequences of using modded assets in dungeons
- plan is to get support from @DubhHerder to migrate custom monsters from DMPRO

### CODE REVIEW

**Comment by ThyWoof** [comment]
it doesn't make much sense to use builders on **DmProEditorContext.cs**. It would simply make the code more verbose.

File: SolastaCommunityExpansion/Models/DmProEditorContext.cs [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+using SolastaModApi.Extensions;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+
+namespace SolastaCommunityExpansion.Models
+{
+    internal static class DmProEditorContext
+    {
+        private const string GUID = "bff53ba4bb694bf5a69b3ae280eec118";
```
Comment (PhilPJL) [inline SolastaCommunityExpansion/Models/DmProEditorContext.cs:11]: maybe change this to
`private static readonly Guid NamespaceGuid = new Guid("...");`
then you eliminate several `new System.Guid(GUID)`

File: SolastaCommunityExpansion/Models/DmProEditorContext.cs [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+
+        private static void UnleashRoomsOnAllEnvironments()
+        {
+            var dbRoomBlueprint = DatabaseRepository.GetDatabase<RoomBlueprint>();
+            var dbEnvironmentDefinition = DatabaseRepository.GetDatabase<EnvironmentDefinition>();
+            var newRooms = new List<RoomBlueprint>();
+
+            foreach (var roomBlueprint in dbRoomBlueprint)
+            {
+                if (roomBlueprint.Category == "EmptyRooms" || roomBlueprint.Category == "FlatRooms")
```
Comment (PhilPJL) [inline SolastaCommunityExpansion/Models/DmProEditorContext.cs:250]: Where(...)

File: SolastaCommunityExpansion/Models/DmProRendererContext.cs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+        private static bool IsFlatRoom(UserRoom userRoom) => userRoom.RoomBlueprint.name.StartsWith(FLAT_ROOM_TAG);
+
+        private static bool IsDynamicFlatRoom(UserRoom userRoom) => IsFlatRoom(userRoom) && int.TryParse(userRoom.RoomBlueprint.name.Substring(FLAT_ROOM_TAG.Length, 2), out int _);
+
+        public static void GetTemplateVegetationMaskArea(WorldLocation worldLocation)
+        {
+            var prefabByReference = worldLocation.GetField<WorldLocation, Dictionary<AssetReference, GameObject>>("prefabByReference");
+
+            foreach (var prefab in prefabByReference.Values)
```
Comment (PhilPJL) [inline SolastaCommunityExpansion/Models/DmProRendererContext.cs:32]: FirstOrDefault

File: SolastaCommunityExpansion/Patches/DungeonMaker/Pro/BaseBlueprintPatcher.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
+﻿using HarmonyLib;
+
+namespace SolastaCommunityExpansion.Patches.DungeonMaker.Pro
+{
+    internal static class BaseBlueprintPatcher
```
Comment (PhilPJL) [inline SolastaCommunityExpansion/Patches/DungeonMaker/Pro/BaseBlueprintPatcher.cs:5]: is this outer class needed?

File: SolastaCommunityExpansion/Patches/DungeonMaker/Pro/UserLocationSettingsModalPatcher.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+using TMPro;
+using static SolastaCommunityExpansion.Models.DmProEditorContext;
+
+namespace SolastaCommunityExpansion.Patches.DungeonMaker.Pro
+{
+    internal static class UserLocationSettingsModalPatcher
+    {
+        // unlocks visual moods across all environments
+        [HarmonyPatch(typeof(UserLocationSettingsModal), "RefreshVisualMoods")]
+        internal static class UserLocationSettingsModalRefreshVisualMoods
```
Comment (PhilPJL) [inline SolastaCommunityExpansion/Patches/DungeonMaker/Pro/UserLocationSettingsModalPatcher.cs:14]: patches not following naming convention elsewhere

**Review by PhilPJL — CHANGES_REQUESTED** [review]
See comments

File: SolastaCommunityExpansion/Models/DmProEditorContext.cs (hunk 1 above)
Comment (ThyWoof) [reply SolastaCommunityExpansion/Models/DmProEditorContext.cs]: fixed

File: SolastaCommunityExpansion/Models/DmProEditorContext.cs (hunk 2 above)
Comment (ThyWoof) [reply SolastaCommunityExpansion/Models/DmProEditorContext.cs]: fixed

File: SolastaCommunityExpansion/Models/DmProRendererContext.cs (hunk 3 above)
Comment (ThyWoof) [reply SolastaCommunityExpansion/Models/DmProRendererContext.cs:32]: Wouldn't we get an unnecessary call to GetComponentInChildren if we use FirstOrDefault? Inside lambda and then outside if return from FirstOrDefault != null?

File: SolastaCommunityExpansion/Patches/DungeonMaker/Pro/BaseBlueprintPatcher.cs (hunk 4 above)
Comment (ThyWoof) [reply SolastaCommunityExpansion/Patches/DungeonMaker/Pro/BaseBlueprintPatcher.cs]: fixed

File: SolastaCommunityExpansion/Patches/DungeonMaker/Pro/UserLocationSettingsModalPatcher.cs (hunk 5 above)
Comment (ThyWoof) [reply SolastaCommunityExpansion/Patches/DungeonMaker/Pro/UserLocationSettingsModalPatcher.cs]: fixed

File: SolastaCommunityExpansion/Models/DmProRendererContext.cs (hunk 3 above)
Comment (PhilPJL) [reply SolastaCommunityExpansion/Models/DmProRendererContext.cs:32]: not sure - leave it

**Review by PhilPJL — APPROVED** [review]

**Pull request merged**